### PR TITLE
feat: library for building predicate parsers

### PR DESCRIPTION
**Pull request by nklaassen**
Took some things I learned while building Login Rules and reviewing `lib/utils/parse`, and implemented a library that can be used to quickly implement new and improved predicate parsers.

Copied from the package comment:

> typical (TYPed predICAte Library) is a library for building better predicate expression parsers faster. It is built on top of [predicate](github.com/gravitational/predicate).
>
> typical helps you (and forces you) to separate parse and evaluation into two distinct stages so that parsing can be cached and evaluation can be fast. Expressions can also be parsed to check their syntax without needing to evaluate them with some specific input.
>
> Functions can be defined by providing implementations accepting and returning values of ordinary types, the library handles the details that enable any function argument to be a subexpression that evaluates to the correct type.
>
> By default, typical provides strong type checking at parse time without any reflection during evaluation. If you define a function that evaluates to type any, you are opting in to evaluation-time type checking everywhere that function is called, but types will still be checked for all other parts of the expression. If you define a function that accepts an interface type it will be called via reflection during evaluation, but interface satisfaction is still checked at parse time.

To see how this will work, in addition to the unit tests here, I've posted:
- the new label expressions parser https://github.com/gravitational/teleport/pull/26176
- conversion of the login rule parser https://github.com/gravitational/teleport.e/pull/1389
- conversion of the parsers in `lib/utils/parse` https://github.com/gravitational/teleport/pull/26313

### CODE REVIEW

**Comment by nklaassen** [comment]
Hey @timothyb89 @gzdunek sorry to throw a 1300 line PR of generic parser code at you, in my defence a bunch of the parser code is pretty repetitive and a good chunk of it is test cases... but let me know if you think it would help for me to chunk it down into smaller PRs and I can give that a shot. Not expecting reviews on a Friday but I'm going to be out until next Thursday so wanted to send this out

**Review by gzdunek — APPROVED** [review]
Looks good, but my experience with Go is limited (and the amount of generic code doesn't help, unfortunately).

>let me know if you think it would help for me to chunk it down into smaller PRs and I can give that a shot. Not expecting reviews on a Friday but I'm going to be out until next Thursday so wanted to send this out

Sorry, I completely missed that message. It seems to me that splitting the code won't help that much, it will probably be difficult to extract independent parts of this function.

File: lib/utils/typical/parser.go [hunk 1]
```diff
@@ -0,0 +1,877 @@
...
+
+func (e ternaryVariadicFuncExpr[TEnv, TArg1, TArg2, TVarArgs, TResult]) Evaluate(env TEnv) (TResult, error) {
+	var nul TResult
+	arg1, err := e.arg1Expr.Evaluate(env)
+	if err != nil {
+		return nul, trace.Wrap(err, "evaluating first argument to function (%s)", e.name)
+	}
+	arg2, err := e.arg2Expr.Evaluate(env)
+	if err != nil {
+		return nul, trace.Wrap(err, "evaluating first argument to function (%s)", e.name)
```
Comment (timothyb89) [inline lib/utils/typical/parser.go:634]:
```suggestion
		return nul, trace.Wrap(err, "evaluating second argument to function (%s)", e.name)
```

**Review by timothyb89 — APPROVED** [review]
This is really cool! I wish we had it when I was futzing with predicate expressions a few months ago, you've solved all the hard problems :slightly_smiling_face:

**Pull request merged**